crates/core/tedge_mapper/src/mapper.rs

```diff
@@ -85,14 +89,121 @@ impl Mapper {
+    /// Reconciliation is done by maintaining an internal journal of all the alarms processed by this mapper,
+    /// which is compared against all the live alarms seen by the mapper on every startup.
+    ///
+    /// All the live alarms are received from tedge/alarms topic on startup.
+    /// Similarly, all the previously processed alarms are received from tedge-internal/alarms topic.
+    /// Reconiciliation detects the difference between these two sets, which are the missed messages.
+    ///
+    /// An alarm that is present in tedge-internal/alarms, but not in tedge/alarms topic
+    /// is assumed to have been cleared while the mapper process was down.
+    /// Similarly, an alarm that is present in tedge/alarms, but not in tedge-internal/alarms topic
+    /// is one that was raised while the mapper process was down.
+    /// An alarm present in both, if their payload is the same, is one that was already processed before the restart
+    /// and hence can be ignored during reconcicilation.
+    async fn reconcile_missed_alarms(&mut self) {
+        let mut pending_alarms_map = HashMap::new();
+        let mut old_alarms_map = HashMap::new();
+
+        let reconciliation_window = tokio::time::sleep(Duration::from_secs(3));
+        tokio::pin!(reconciliation_window); // Explicit pinning is required to use this sleep Future in a select! loop
+
+        loop {
+            tokio::select! {
+                // When the reconciliation window times out, break out of the select loop to stop caching alarm messages
+                _ = &mut reconciliation_window => {
+                    break;
+                }
+
+                // During the reconciliation window, cache all the alarm messages received from tedge/alarms and tedge-internal alarms topics.
+                // This window gives the mapper ample time to receive all the retained messages from both the topics.
+                Some(message) = self.input.next() => {
+                    if message.topic.name.starts_with("tedge/alarms") {
+                        let alarm_id = message
+                            .topic
+                            .name
+                            .strip_prefix("tedge/alarms/")
+                            .expect("Expected tedge/alarms prefix")
+                            .to_string();
+                        pending_alarms_map
+                            .insert(alarm_id.clone(), message);
+                    } else if message.topic.name.starts_with("tedge-internal/alarms") {
+                        let alarm_id = message
+                            .topic
+                            .name
+                            .strip_prefix("tedge-internal/alarms/")
+                            .expect("Expected tedge-internal/alarms prefix")
+                            .to_string();
+                        old_alarms_map
+                            .insert(alarm_id, message);
+                    }
+                }
+            }
+        }
```

Comment:
I would rather use `tokio::timeout` to drive this instead of manually pushing the `select` loop, 

```suggestion
        let _ = tokio::time::timeout(Duration::from_secs(3), async {
            while let Some(message) = self.input.next().await {
                if message.topic.name.starts_with("tedge/alarms") {
                    let alarm_id = message
                        .topic
                        .name
                        .strip_prefix("tedge/alarms/")
                        .expect("Expected tedge/alarms prefix")
                        .to_string();
                    pending_alarms_map.insert(alarm_id.clone(), message);
                } else if message.topic.name.starts_with("tedge-internal/alarms") {
                    let alarm_id = message
                        .topic
                        .name
                        .strip_prefix("tedge-internal/alarms/")
                        .expect("Expected tedge-internal/alarms prefix")
                        .to_string();
                    old_alarms_map.insert(alarm_id, message);
                }
            }
        })
        .await;
```

Reply:
This might work, but I don't know if the usage of `timeout` is semantically correct here. The documentation of `tokio::timeout` states that it will cancel the underlying future of the task doesn't finish within the given timeframe. So, it kinda indicates a failure. But as per our logic, what we need is more of a **window** during which a task is performed. So, for a reader, using a timeout here might be a bit confusing, I feel. But I agree that it makes the code much simpler. So, not fully sure. @didier-wenzek WDYT?

BTW, I don't know what the implication of that future cancellation is. Hope there is none as our task just populates some maps and doesn't do anything blocking.

Reply:
That's good point, I think that first the work proposed by Didier should be done to move this part to the converter this part my not even be required at all.

Reply:
I think that the two alternatives are equivalent regarding cancellation of the message loop. Hence, using a timeout will simplify the code.